Client/Frontend/Browser/BrowserViewController/OpenSearch/BrowserViewController+OpenSearch.swift

```diff
@@ -0,0 +1,212 @@
+        case title = "title"
+    }
+}
+
+// MARK: - OpenSearch Browser Extension
+
+extension BrowserViewController {
+    
+    /// Adding Toolbar button over the keyboard for adding Open Search Engine
+    /// - Parameter webView: webview triggered open seach engine
+    func evaluateWebsiteSupportOpenSearchEngine(_ webView: WKWebView) {
+        let script = """
+                        var link = document.querySelector("link[type='application/opensearchdescription+xml']")
+                        var dict = {
+                            href : link.getAttribute("href"),
+                            title : link.getAttribute("title")
+                        };
+                        JSON.stringify(dict)
+                     """
+        
+        webView.evaluateJavaScript(script) { (result, _) in
+            guard let htmlStr = result as? String,
+                  let data: Data = htmlStr.data(using: .utf8) else { return }
+                        
+            do {
+                let openSearchReference = try JSONDecoder().decode(OpenSearchReference.self, from: data)
+                self.updateAddOpenSearchEngine(webView, referenceObject: openSearchReference)
+            } catch {
+                log.error(error.localizedDescription)
+            }
+        }
+    }
+    
+    private func updateAddOpenSearchEngine(_ webView: WKWebView, referenceObject: OpenSearchReference) {
+        // Add Reference Object as Open Search Engine
+        openSearchEngine = referenceObject
+        
+        // Open Search guidlines requires Title to be same as Short Name but it is not enforced,
+        // thus in case of yahoo.com the title is 'Yahoo Search' and Shortname is 'Yahoo'
+        // Instead we are checking referenceURL match to determine searchEngine is added or not
+        
+        let matches = self.profile.searchEngines.orderedEngines.filter {$0.referenceURL == referenceObject.reference}
+        
+        if !matches.isEmpty {
+            self.customSearchEngineButton.state = .disabled
+        } else {
+            self.customSearchEngineButton.state = .enabled
+        }
+        
+        /*
+         This is how we access hidden views in the WKContentView
+         Using the public headers we can find the keyboard accessoryView which is not usually available.
+         Specific values here are from the WKContentView headers.
+         https://github.com/JaviSoto/iOS9-Runtime-Headers/blob/master/Frameworks/WebKit.framework/WKContentView.h
+         */
+        guard let webContentView = UIView.findSubViewWithFirstResponder(webView) else {
+            /*
+             In some cases the URL bar can trigger the keyboard notification. In that case the webview isnt the first responder
+             and a search button should not be added.
+             */
+            return
+        }
+        
+        guard let input = webContentView.perform(#selector(getter: UIResponder.inputAccessoryView)),
```

Comment:
You should definitely check if that selector exists first before executing it because if this changes on a different version of iOS, we'll get a crash. Happened to me with videos in WKWebView.

Reply:
Yeah this part was from exactly how Firefox used to add a toolbar button on top of keyboard.
However you are totally right I will add selector check and also scrambling.